### PR TITLE
NXDRIVE-1926: [Direct Transfer] Display global transfer metrics

### PR DESCRIPTION
**Pull request by rom1win**
WIP

### CODE REVIEW

File: nxdrive/data/i18n/i18n.json [hunk 1]
```diff
@@ -85,7 +85,8 @@
     "DIRECT_EDIT_UNLOCK_ERROR": "Unlocking issue",
     "DIRECT_EDIT_UNLOCK_ERROR_DESCRIPTION": "The document „%1“ was not unlocked",
     "DIRECT_EDIT_UPDATED_FILE": "The file „%1“ has been updated",
-    "DIRECT_EDIT_VERSION": "You are trying to perform a Direct Edit on version %1 of „%2“. It is not the current version of the document, so it is readonly and cannot be edited.",
+    "DIRECT_EDIT_VERSION": "You are trying to perform a DirectEdit on version %1 of „%2“. It is not the current version of the document, so it is readonly and cannot be edited.",
```
Comment (BoboTiG) [inline nxdrive/data/i18n/i18n.json:88]: Nope :)

File: nxdrive/data/i18n/i18n.json [hunk 2]
```diff
@@ -86,12 +86,14 @@
...
     "DIRECT_EDIT_UPDATED_FILE": "The file „%1“ has been updated",
     "DIRECT_EDIT_VERSION": "You are trying to perform a Direct Edit on version %1 of „%2“. It is not the current version of the document, so it is readonly and cannot be edited.",
+    "DIRECT_TRANSFER_DETAILS": "%1 of %2",
     "DIRECT_TRANSFER_END": "Transfer done: „%1“",
     "DIRECT_TRANSFER_ERROR": "Transfer error: „%1“",
     "DIRECT_TRANSFER_NO_ACCOUNT": "Cannot use the Direct Transfer feature with no account, aborting.",
     "DIRECT_TRANSFER_NOT_ALLOWED": "Direct Transfer of „%1“ is not allowed for synced files.",
     "DIRECT_TRANSFER_NOT_ENABLED": "The Direct Transfer feature is not enabled.",
     "DIRECT_TRANSFER_NOT_POSSIBLE": "The Direct Transfer feature is not usable right now.<br>Please try again later.",
+    "DIRECT_TRANSFER_SEND": "Sending files to ",
```
Comment (BoboTiG) [inline nxdrive/data/i18n/i18n.json:96]:
```suggestion
    "DIRECT_TRANSFER_SEND": "Sending files to <a href='%1'>%2</a>",
```

File: nxdrive/gui/view.py [hunk 3]
```diff
@@ -256,6 +256,131 @@ def flags(self, index: QModelIndex) -> Qt.ItemFlags:
...
+        for i, item in enumerate(self.transfers):
+            if item["name"] != action["name"]:
+                continue
+            idx = self.createIndex(i, 0)
+
+            if action["action_type"] in ("Linking", "Verification"):
+                # Disable the speed to not show the speed at the final step
+                item["speed"] = 0
+            else:
+                item["speed"] = action["speed"]
```
Comment (BoboTiG) [inline nxdrive/gui/view.py:373]: Can be removed, right? No speed details are printed.

File: nxdrive/data/qml/TransferElem.qml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                icon: paused ? MdiFont.Icon.play : MdiFont.Icon.pause
+                tooltip: qsTr(paused ? "RESUME" : "SUSPEND") + tl.tr
+                onClicked: {
+                    // engine is set for Direct Edit transfers only
+                    var engine_uid = engine || accountSelect.getRole("uid")
+                    if (paused) {
+                        api.resume_transfer("upload", engine_uid, uid)
+                    } else {
+                        api.pause_transfer("upload", engine_uid, uid, progress)
+                    }
```
Comment (BoboTiG) [inline nxdrive/data/qml/TransferElem.qml:71]:
```suggestion
                    if (paused) {
                        api.resume_transfer("upload", engine, uid)
                    } else {
                        api.pause_transfer("upload", engine, uid, progress)
                    }
```

File: nxdrive/gui/api.py [hunk 5]
```diff
@@ -201,6 +202,34 @@ def get_transfers(self) -> List[Dict[str, Any]]:
 
         return result
 
+    def upload_is_direct_transfer(self, path, engine):
```
Comment (BoboTiG) [inline nxdrive/gui/api.py:205]: Maybe a `upload.is_direct_transfer` is preferable to prefevent such code.

File: nxdrive/gui/api.py [hunk 6]
```diff
@@ -201,6 +202,34 @@ def get_transfers(self) -> List[Dict[str, Any]]:
...
+        if engine.local_folder not in path.parents:
+            return True
+        elif not engine.dao.get_state_from_local(path.relative_to(engine.local_folder)):
+            return True
+        return False
+
+    @pyqtSlot(result=list)
+    def get_direct_transfers(self) -> List[Dict[str, Any]]:
```
Comment (BoboTiG) [inline nxdrive/gui/api.py:213]: Missing doctstring + changelog.
```suggestion
    def get_direct_transfer_items(self) -> List[Dict[str, Any]]:
```

File: nxdrive/gui/api.py [hunk 7]
```diff
@@ -315,7 +344,19 @@ def open_direct_transfer(self, uid: str) -> None:
...
+                for dt in list(engine.dao.get_uploads())
+                if self.upload_is_direct_transfer(dt.path, engine)
+            ]
+            if direct_transfers:
+                self.application.show_direct_transfer_window(engine.uid)
+            else:
+                self.application.show_server_folders(engine, None)
+
+    @pyqtSlot()
+    def close_direct_transfer(self) -> None:
```
Comment (BoboTiG) [inline nxdrive/gui/api.py:358]: Missing doctstring + types + changelog.
```suggestion
    def close_direct_transfer_windows(self) -> None:
```

File: nxdrive/gui/application.py [hunk 8]
```diff
@@ -313,6 +324,7 @@ def action_progressing(self, action: Action) -> None:
             log.warning(f"An action is needed, got {action!r}")
             return
         self.transfer_model.set_progress(action.export())
+        self.direct_transfer_model.set_progress(action.export())
```
Comment (BoboTiG) [inline nxdrive/gui/application.py:327]: We should have a way to make the difference. `action.export()` should be stored into a variable and determine if this is a DT or not to not call `.set_progress` on all models.

File: nxdrive/gui/application.py [hunk 9]
```diff
@@ -685,6 +698,27 @@ def show_conflicts_resolution(self, engine: Engine) -> None:
...
+        self._window_root(self.direct_transfer_window).setEngine.emit(engine_uid)
+        screen = self.direct_transfer_window.screen()
+        height = screen.size().height()
+        width = screen.size().width()
+        self.direct_transfer_window.setX(
+            (width / 2) - (self.direct_transfer_window.minimumWidth() / 2)
+        )
+        self.direct_transfer_window.setY(
+            (height / 2) - (self.direct_transfer_window.minimumHeight() / 2)
+        )
```
Comment (BoboTiG) [inline nxdrive/gui/application.py:713]: - Add blank lines before and after the code block.
- Add a comment like "# Center the window on the screen"

File: nxdrive/gui/application.py [hunk 10]
```diff
@@ -685,6 +698,27 @@ def show_conflicts_resolution(self, engine: Engine) -> None:
...
+        self.direct_transfer_window.show()
+        self.direct_transfer_window.requestActivate()
+
+    @pyqtSlot(str)
```
Comment (BoboTiG) [inline nxdrive/gui/application.py:717]: Should be removed.

File: nxdrive/gui/application.py [hunk 11]
```diff
@@ -757,9 +791,16 @@ def show_server_folders(self, engine: Engine, path: Optional[Path]) -> None:
             self.filters_dlg = None
 
         self.filters_dlg = FoldersDialog(self, engine, path)
-        self.filters_dlg.destroyed.connect(self.destroyed_filters_dialog)
+        self.filters_dlg.destroyed.connect(self.destroyed_server_folders)
         self.filters_dlg.show()
 
+    @pyqtSlot()
+    def destroyed_server_folders(self) -> None:
```
Comment (BoboTiG) [inline nxdrive/gui/application.py:798]: Missing doctstring + changelog.

File: nxdrive/gui/application.py [hunk 12]
```diff
@@ -1529,7 +1570,15 @@ def update_status(self, engine: Engine) -> None:
 
     @pyqtSlot()
     def refresh_transfers(self) -> None:
-        transfers = self.api.get_transfers()
```
Comment (BoboTiG) [inline nxdrive/gui/application.py:1532]: Nope, nope, nope :)

Add a specific method for Direct Transfer items. Do not interfere here.

File: nxdrive/gui/view.py [hunk 13]
```diff
@@ -256,6 +256,131 @@ def flags(self, index: QModelIndex) -> Qt.ItemFlags:
         return Qt.ItemIsEditable | Qt.ItemIsEnabled | Qt.ItemIsSelectable
 
 
+class DirectTransferModel(QAbstractListModel):
+    fileChanged = pyqtSignal()
+    emptyDt = pyqtSignal()
```
Comment (BoboTiG) [inline nxdrive/gui/view.py:261]:
```suggestion
    noItems = pyqtSignal()
```

File: nxdrive/gui/view.py [hunk 14]
```diff
@@ -256,6 +256,131 @@ def flags(self, index: QModelIndex) -> Qt.ItemFlags:
...
+    STATUS = Qt.UserRole + 3
+    PROGRESS = Qt.UserRole + 4
+    ENGINE = Qt.UserRole + 5
+    FINALIZING = Qt.UserRole + 6
+    PROGRESS_METRICS = Qt.UserRole + 7
+
+    def __init__(self, translate: Callable, parent: QObject = None) -> None:
+        super().__init__(parent)
+        self.tr = translate
+        self.transfers: List[Dict[str, Any]] = []
```
Comment (BoboTiG) [inline nxdrive/gui/view.py:274]:
```suggestion
        self.items: List[Dict[str, Any]] = []
```

File: nxdrive/gui/view.py [hunk 15]
```diff
@@ -256,6 +256,131 @@ def flags(self, index: QModelIndex) -> Qt.ItemFlags:
...
+        count = self.rowCount()
+        if count == 0:
+            self.emptyDt.emit()
+        return count
+
+    @pyqtProperty(list, notify=fileChanged)
+    def destination(self) -> List[str]:
+        return self.get_dest()
+
+    def set_transfers(
```
Comment (BoboTiG) [inline nxdrive/gui/view.py:317]:
```suggestion
    def set_items(
```

File: nxdrive/client/uploader/direct_transfer.py [hunk 16]
```diff
@@ -141,8 +141,10 @@ def upload(
         # Save the remote document's UID into the file xattrs, in case next steps fails
         if remote_ref:
             LocalClient.set_path_remote_id(file_path, remote_ref, name="remote")
-
-        # Upload the blob and use the FileManager importer to create the document
```
[the remark this answers is not in the document]
Comment (rom1win) [reply nxdrive/client/uploader/direct_transfer.py:145]: mistake

File: nxdrive/gui/api.py (hunk 5 above)
Comment (rom1win) [reply nxdrive/gui/api.py:205]: the same than `is_direct_edit` ?

File: nxdrive/data/qml/TransferElem.qml [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    ColumnLayout {
+        id: transfer
+        anchors.fill: parent
+        anchors.centerIn: parent
+        spacing: 3
+
+        // Progression: transferred data and remote folder
+        ScaledText {
+            property string pretty_progress: "[" + Math.floor(progress || 0) + "%] "
+            text: pretty_progress + qsTr("DIRECT_TRANSFER_DETAILS").arg(progress_metrics[0]).arg(progress_metrics[1]) + tl.tr
```
[the remark this answers is not in the document]
Comment (rom1win) [reply nxdrive/data/qml/TransferElem.qml:23]: format doesn't seems to work

File: nxdrive/gui/api.py (hunk 5 above)
Comment (BoboTiG) [reply nxdrive/gui/api.py:205]: Yes.

File: nxdrive/gui/application.py [hunk 18]
```diff
@@ -757,9 +792,20 @@ def show_server_folders(self, engine: Engine, path: Optional[Path]) -> None:
...
         self.filters_dlg = FoldersDialog(self, engine, path)
-        self.filters_dlg.destroyed.connect(self.destroyed_filters_dialog)
+        self.filters_dlg.destroyed.connect(self.destroyed_server_folders)
         self.filters_dlg.show()
 
+    @pyqtSlot()
+    def destroyed_server_folders(self) -> None:
+        """
+        Called when the server folders selection dialog is destroyed.
+        Show the Direct Transfer Window if a folder is selected.
```
Comment (BoboTiG) [inline nxdrive/gui/application.py:802]:
```suggestion
        Show the Direct Transfer window if a folder is selected.
```

File: nxdrive/gui/view.py [hunk 19]
```diff
@@ -256,6 +256,125 @@ def flags(self, index: QModelIndex) -> Qt.ItemFlags:
...
+        self.beginInsertRows(parent, 0, len(transfers) - 1)
+        self.items.extend(transfers)
+        self.endInsertRows()
+
+        self.fileChanged.emit()
+
+    def get_progress(self, row: Dict[str, Any]) -> str:
+        """Return formatted infos list on direct transfer progression.
+        E.g: ["10.0Mib", "42.0Mib", "remote_path"]
+        """
```
Comment (BoboTiG) [inline nxdrive/gui/view.py:333]:
```suggestion
    def get_progress(self, row: Dict[str, Any]) -> Tuple[str, str]:
        """Return a tuple (transferred_size, transfer_size) nicely formatted."""
        """
```

File: nxdrive/data/qml/TransferElem.qml (hunk 17 above)
[the remark this answers is not in the document]
Comment (BoboTiG) [reply nxdrive/data/qml/TransferElem.qml]: What is the result? Here is the revelant doc: https://doc.qt.io/qt-5/qstring.html#arg-9

File: nxdrive/gui/application.py (hunk 12 above)
Comment (BoboTiG) [reply nxdrive/gui/application.py:1532]: Finally, you have to use it but the code logic must be adapter. As of now, this is not acceptable :)

But you can make use of the new `upload.is_direct_transfer` attribute now.

File: nxdrive/client/uploader/__init__.py [hunk 20]
```diff
@@ -163,7 +163,8 @@ def upload_chunks(
 
             engine_uid = kwargs.pop("engine_uid", None)
             is_direct_edit = kwargs.pop("is_direct_edit", False)
-
+            is_direct_transfer = kwargs.pop("is_direct_transfer", False)
+            remote_ref = kwargs.pop("remote_ref", "")
             # Set those attributes as FileBlob does not have them
```
Comment (BoboTiG) [inline nxdrive/client/uploader/__init__.py:168]:
```suggestion

            # Set those attributes as FileBlob does not have them
```

File: nxdrive/engine/activity.py [hunk 21]
```diff
@@ -124,6 +124,9 @@ def __init__(
         # Number of chunks transferred since the last speed computation
         self.transferred_chunks = 0
 
+        # Used to kwnow if the file is a Direct Transfer
```
Comment (BoboTiG) [inline nxdrive/engine/activity.py:127]:
```suggestion
        # Used to know if the file is a Direct Transfer item
```

**Pull request merged**